lib/trino-parquet/src/main/java/io/trino/parquet/BloomFilterStore.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.parquet.hadoop.metadata.ColumnChunkMetaData;
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+import org.apache.parquet.io.ParquetDecodingException;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Verify.verify;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.apache.parquet.column.values.bloomfilter.BlockSplitBloomFilter.UPPER_BOUND_BYTES;
+
+public class BloomFilterStore
+{
+    // since bloomfilter header is relatively small(18 bytes when testing) we can read in a larger buffer(BlockSplitBloomFilter.HEADER_SIZE*4 in this case)
+    // and get actual bytes used when deserializing header in order to calculate the correct offset for bloomfilter data.
+    private static final int MAX_HEADER_LENGTH = BlockSplitBloomFilter.HEADER_SIZE * 4;
+
+    private final ParquetDataSource dataSource;
+    private final Map<ColumnPath, Long> bloomFilterOffsets;
+
+    public BloomFilterStore(ParquetDataSource dataSource, BlockMetaData block, Set<ColumnPath> columnsFiltered)
+    {
+        this.dataSource = requireNonNull(dataSource, "dataSource is null");
+        requireNonNull(block, "block is null");
+        requireNonNull(columnsFiltered, "columnsFiltered is null");
+
+        ImmutableMap.Builder<ColumnPath, Long> bloomFilterOffsetBuilder = ImmutableMap.builder();
+        for (ColumnChunkMetaData column : block.getColumns()) {
+            ColumnPath path = column.getPath();
+            if (hasBloomFilter(column) && columnsFiltered.contains(path)) {
+                bloomFilterOffsetBuilder.put(path, column.getBloomFilterOffset());
+            }
+        }
+        this.bloomFilterOffsets = bloomFilterOffsetBuilder.buildOrThrow();
+    }
+
+    public Optional<BloomFilter> getBloomFilter(ColumnPath columnPath)
+    {
+        BloomFilterHeader bloomFilterHeader;
+        long bloomFilterDataOffset;
+        try {
+            Long columnBloomFilterOffset = bloomFilterOffsets.get(columnPath);
+            if (columnBloomFilterOffset == null) {
+                return Optional.empty();
+            }
+            Slice headerSlice = dataSource.readFully(columnBloomFilterOffset, MAX_HEADER_LENGTH);
+            BasicSliceInput input = headerSlice.getInput();
+            bloomFilterHeader = Util.readBloomFilterHeader(input);
+            bloomFilterDataOffset = columnBloomFilterOffset + input.position();
+        }
+        catch (IOException exception) {
+            throw new UncheckedIOException("Failed to read Bloom filter header", exception);
+        }
+
+        if (!bloomFilterSupported(columnPath, bloomFilterHeader)) {
+            return Optional.empty();
+        }
+
+        try {
+            Slice data = dataSource.readFully(bloomFilterDataOffset, bloomFilterHeader.getNumBytes());
```

Comment:
let's avoid using variable name such as "data", "input" as the reader cannot tells if the code is reading the header or actual data.